### PR TITLE
Add better errors for empty-spec.

### PR DESCRIPTION
**Pull request by brendandburns**
Closes https://github.com/GoogleCloudPlatform/kubernetes/issues/9012

### CODE REVIEW

**Comment by quinton-hoole** [comment]
LGTM, but what about other API types?  This only covers pods, services and replication controllers.  Also, it would be good if we could factor this out generically rather than have so many duplicate pieces of essentially the same code.

**Comment by quinton-hoole** [comment]
cc thockin FYI, as he opened #9343

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -943,8 +943,11 @@ func validateImagePullSecrets(imagePullSecrets []api.LocalObjectReference) errs.
 func ValidatePod(pod *api.Pod) errs.ValidationErrorList {
 	allErrs := errs.ValidationErrorList{}
 	allErrs = append(allErrs, ValidateObjectMeta(&pod.ObjectMeta, true, ValidatePodName).Prefix("metadata")...)
-	allErrs = append(allErrs, ValidatePodSpec(&pod.Spec).Prefix("spec")...)
-
+	if reflect.DeepEqual(&api.PodSpec{}, &pod.Spec) {
```
Comment (mikedanese) [inline pkg/api/validation/validation.go]: At this point hasn't the pod.Spec already been defaulted?

**Comment by mikedanese** [comment]
I discussed this with @thockin today. IIUC, I think it's actually pretty involved as defaults have already been applied in validation. It circled back to the "we shouldn't be converting in clients" discussion.

**Comment by thockin** [comment]
I think it is more than just not converting in clients. Validation
necessarily happens after defaulting, and we no easy way to tell if spec
was provided or not after we convert from JSON.

We could receive spec as a pointer (in versioned) and not set defaults if
it is nil. Sort of tedious and bound to hit nil-derefs. Worse, I can
plausibly imagine objects with no required spec fields at create time,
meaning a non-provided spec is OK.

Maybe I just wrong, and this is the best error we can do for now.

**Comment by quinton-hoole** [comment]
Are we still wanting to merge this?  Review seems inconclusive?

**Comment by quinton-hoole** [comment]
@mikedanese @thockin @brendandburns ^^^

**Comment by brendandburns** [comment]
I will test this in a live cluster and see if I get a better error message, if not I'll fix this PR.

**Comment by thockin** [comment]
yeah, I think this is dead.  We might do better at errors, somehow, but I don't think this is the right way.

**Pull request closed without merging**